Public/Src/Engine/Processes/SharedOpaqueOutputHelper.cs

```diff
@@ -10,31 +11,140 @@
+                if (writeAttributesDenied)
+                {
+                    FileUtilities.SetFileAccessControl(expandedPath, FileSystemRights.WriteAttributes | FileSystemRights.WriteExtendedAttributes, allow: true);
+                }
+
+                try
+                {
+                    // Only the creation time is used to identify a file as the output of a shared opaque
+                    FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                }
+                catch (BuildXLException ex)
+                {
+                    // On (unsafe) double writes, a race can occur, we give the other contender a second chance
+                    if (IsSharedOpaqueOutput(expandedPath))
+                    {
+                        return;
+                    }
+
+                    // Since these files should be just created outputs, this shouldn't happen and we bail out hard.
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for writing."), ex);
+                }
+                finally
+                {
+                    // Restore the attributes as they were originally set
+                    if (writeAttributesDenied)
+                    {
+                        FileUtilities.SetFileAccessControl(expandedPath, FileSystemRights.WriteAttributes | FileSystemRights.WriteExtendedAttributes, allow: false);
+                    }
+                }
+            }
+
+            /// <summary>
+            /// Checks if the given path is an output under a shared opaque by verifying whether <see cref="WellKnownTimestamps.OutputInSharedOpaqueTimestamp"/> is the creation time of the file
+            /// </summary>
+            /// <remarks>
+            /// If the given path is a directory, it is always considered part of a shared opaque
+            /// </remarks>
+            public static bool IsSharedOpaqueOutput(string expandedPath)
             {
-                // Only the creation time is used to identify a file as the output of a shared opaque
-                FileUtilities.SetFileTimestamps(expandedPath, new FileTimestamps(WellKnownTimestamps.OutputInSharedOpaqueTimestamp));
+                try
+                {
+                    var creationTime = FileUtilities.GetFileTimestamps(expandedPath).CreationTime;
+                    return creationTime == WellKnownTimestamps.OutputInSharedOpaqueTimestamp;
+                }
+                catch (BuildXLException ex)
+                {
+                    throw new BuildXLException(I($"Failed to open output file '{expandedPath}' for reading."), ex);
+                }
             }
-            catch (BuildXLException ex)
+        }
+
+        private static unsafe class Unix
+        {
+            private const string MY_XATTR_NAME = "com.microsoft.buildxl:shared_opaque_output";
+            private const long MY_XATTR_VALUE = 42;

```

Comment:
> 42 [](start = 47, length = 3)

I assume this is arbitrary, right? Maybe adding a comment stating that?

Reply:
Yes, completely arbitrary; will add a comment.